cloudinit/sources/helpers/vultr.py

```diff
@@ -0,0 +1,337 @@
+# Author: Eric Benner <ebenner@vultr.com>
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+import json
+import os
+import copy
+import re
+
+from cloudinit import log as log
+from cloudinit import url_helper
+from cloudinit import dmi
+from cloudinit import util
+from cloudinit import net
+from cloudinit import subp
+from cloudinit.net.dhcp import EphemeralDHCPv4, NoDHCPLeaseError
+
+# Get logger
+LOGGER = log.getLogger(__name__)
+
+# Cache
+MAC_TO_NICS = None
+METADATA = None
+EHP = None
+
+
+def bring_up_interface(connectivity_url=None):
+    global EHP
+
+    # If for whatever reason this is up, bail
+    if EHP is not None:
+        return
+
+    # Make sure its not up already
+    if net.has_url_connectivity(connectivity_url):
+        return
+
+    # Bring up interface in local
+    try:
+        EHP = EphemeralDHCPv4(net.find_fallback_nic())
+        EHP.obtain_lease()
+    except (NoDHCPLeaseError) as exc:
+        LOGGER.error("DHCP failed, cannot continue. Exception: %s",
+                     exc)
+        raise
+
+
+# Close EphermalDHCP so its not left open
+def close_ephermeral():
+    global EHP
+
+    # No action if its not open
+    if EHP is None:
+        return
+
+    EHP.clean_network()
+
+    # Cleanup
+    EHP = None
+
+
+# Cache the metadata for optimization
+def get_metadata(params):
+    global METADATA
+
+    if not METADATA:
+        # Bring up interface in local
```

Comment:
use a context for this rather htan the global EHP state.

I think you're basically just doing:
```python
with EphemeralDHCPv4(connectivity_url=params['url']):
    v1 = fetch_metadata(params)
```

and that should take care of EPH entirely.